backend/cms/templates/registration/login_mfa.html

```diff
@@ -0,0 +1,141 @@
+        let {challenge, allowCredentials} = credentialRequestOptionsFromServer;
+
+        challenge = Uint8Array.from(
+            atob(challenge), c => c.charCodeAt(0));
+
+        allowCredentials = allowCredentials.map(credentialDescriptor => {
+            let {id} = credentialDescriptor;
+            id = id.replace(/\_/g, "/").replace(/\-/g, "+");
+            id = Uint8Array.from(atob(id), c => c.charCodeAt(0));
+            return Object.assign({}, credentialDescriptor, {id});
+        });
+
+        const transformedCredentialRequestOptions = Object.assign(
+            {},
+            credentialRequestOptionsFromServer,
+            {challenge, allowCredentials});
+
+        return transformedCredentialRequestOptions;
+    };
+
+    const transformAssertionForServer = (newAssertion) => {
+        const authData = new Uint8Array(newAssertion.response.authenticatorData);
+        const clientDataJSON = new Uint8Array(newAssertion.response.clientDataJSON);
+        const rawId = new Uint8Array(newAssertion.rawId);
+        const sig = new Uint8Array(newAssertion.response.signature);
+        const assertionClientExtensions = newAssertion.getClientExtensionResults();
+
+        return {
+            id: newAssertion.id,
+            rawId: b64enc(rawId),
+            type: newAssertion.type,
+            authData: b64RawEnc(authData),
+            clientData: b64RawEnc(clientDataJSON),
+            signature: hexEncode(sig),
+            assertionClientExtensions: JSON.stringify(assertionClientExtensions)
+        };
+    };
+
+    window.addEventListener('load', async () => {
+        try {
+            const webauthn_assert = await(await fetch('{% url "login_mfa_assert" %}')).json();
+            
+            const transformedCredentialRequestOptions = transformCredentialRequestOptions(
+                webauthn_assert);
+
+            // request the authenticator to create an assertion signature using the
+            // credential private key
+            let assertion;
+            assertion = await navigator.credentials.get({
+                publicKey: transformedCredentialRequestOptions,
+            });
+
+            // we now have an authentication assertion! encode the byte arrays contained
+            // in the assertion data as strings for posting to the server
+            const transformedAssertionForServer = transformAssertionForServer(assertion);
+
+            const result = await fetch("{% url 'login_mfa_verify' %}", {
+                method: 'POST',
+                headers: {
+                    'Content-Type': 'application/json',
+                    'X-CSRFToken': window.django_csrf_token
+                },
+                body: JSON.stringify(transformedAssertionForServer)
+            });
+            const data = await result.json();
+            if(data.success) {
+                location.href = "/"
+            } else {
+                document.querySelector('.auth-error').classList.remove('hidden');
```

Comment:
If you want to, you can use [umbrellaJS](https://umbrellajs.com/documentation), but vanilla JS is just fine :wink:

Reply:
`umbrellajs` seems to always returns an array, but I want to query only one element when using `querySelector` (so I will keep the original code)

Reply:
`querySelector` always returns the first matched element, so to get the same functionality in umbrellajs, you'd either use an id-selector instead of classes `u('#auth-error').removeClass('hidden')` or explicitly choose the first element `u('.auth-error').first().removeClass('hidden')`. But that's just a side note, please don't rewrite the js-code, I mean this library should make life easier and not cause pointless work :grin: